### DISCUSSION TITLE
how do I add particle-mesh collision to the DEM example?

### DISCUSSION THREAD

**Discussion by jlinBE** [NVIDIA/warp#22] [Q&A] [answered]
I'm asking about example_dem.py.  Particles can currently bounce off the ground plane, but there's no support for mesh collision.  How can I add it?

I tried lifting the code from the PBD-based mesh examples, but it didn't work very well.  The mesh is in the right place but the particles just bounce off all over the place.  I'm guessing I need to use contact_force() like in the ground contact?

I'm just interested in bouncing spheres off a static mesh, at this point.  Doesn't need to deform.  Anyone have a working example or can point me towards the right papers?

**Comment by cadop** [comment]
Just a note, if you change the "Category" to Q&A, it lets you have the 'select an answer' feature.

**Reply by jlinBE** [reply]
good to know - changed.

**Comment by jlinBE** [comment, marked as answer]
I figured it out:

	#mesh contact
	face_index = int(0)
	face_u = float(0.0)
	face_v = float(0.0)
	sign = float(0.0)
	max_dist = radius
	
	if (wp.mesh_query_point(mesh, x, max_dist, sign, face_index, face_u, face_v)):
		p = wp.mesh_eval_position(mesh, face_index, face_u, face_v)
		n = x - p
		d = wp.length(n)
		err = d - radius*2.0
		if (err < cohesion_ground):
			n = n/d
			f = f + contact_force(n, v, err, k_contact, k_damp, 100.0, 0.5)
